### PR TITLE
colored nodes in graph

### PR DESCRIPTION
**Pull request by ghost**
I hope this is all right
It could probably be done modular and configurable, but this is what I have for now

### CODE REVIEW

**Comment by jplitza** [comment]
I am missing an explanation. What does this do? Why is it useful/necessary?

**Comment by ghost** [comment]
All nodes with a VPN connection are marked blue, all nodes without a mesh connection are marked orange. You can see immediantly, when a VPN link is lost, because someone unplugged ethernet (like you could see it in ffmap-d3).
This was requested by many users.

**Comment by lost-tty** [comment]
This should evaluate the fastd status from statistics (checking whether fastd is enabled and there are active peers) instead of relying on the graph.

**Comment by ghost** [comment]
I saw that while looking at the graph with the example config, which doesn't have servers in the data, but has a p2p vpn connection. I will do this right now. Thanks for the feedback.

**Comment by ghost** [comment]
So where should meshviewer get the information whether there are active peers? Is it in the nodeinfo or statistics?

**Comment by lost-tty** [comment]
It's in statistics.

**Comment by ghost** [comment]
```
  "statistics": {
    "clients": 0,
    "gateway": "04:be:4a:bc:0d:40",
    "loadavg": 0.24,
    "memory_usage": 0.17188538863384517,
    "rootfs_usage": 0.066071428571429,
    "uptime": 704264.75,
    "traffic": {
      ... blah ...
    }
  },
```

I can't find it

**Comment by lost-tty** [comment]
Ah, it may be that it's not forwarded by ffmap-backend yet.

**Comment by ghost** [comment]
So do we increase the nodes.json version number for that? Maybe you can apply the changes to the backend? I am even worse in python.

**Comment by ghost** [comment]
This works in combination with @t-8ch s pr for the backend.

**Comment by lost-tty** [comment]
What are "hotspotNodes"?

**Comment by ghost** [comment]
These are nodes which don't have any wifi mesh connections. This is inspired by KBU's old ffmap-d3 fork

**Comment by lost-tty** [comment]
I don't think that's necessary. They are already easily discernable by the absense of any neighbours.

File: lib/forcegraph.js [hunk 1]
```diff
@@ -341,21 +345,37 @@ define(["d3"], function (d3) {
...
+        ctx.drawImage(node, scale * r * d.x - node.width / 2, scale * r * d.y - node.height / 2)
+      })
+
       ctx.restore()
 
       // -- draw clients --
       ctx.save()
       ctx.beginPath()
-      nodes.filter(visibleNodes).forEach(function (d) {
+      meshNodes.concat(uplinkNodes).concat(hotspotNodes).filter(visibleNodes).forEach(function (d) {
```
Comment (lost-tty) [inline lib/forcegraph.js]: It may make sense to keep something like `allNodes` around to avoid concatting the arrays again and again.

**Comment by ghost** [comment]
This is used in other features i.e. a button to hide these nodes. We can also concat them with the mesh nodes before drawing so that they have the same colors.

**Comment by lost-tty** [comment]
Can we rescope this as "show nodes with uplink in a different color"?

**Comment by ghost** [comment]
better?

**Comment by ghost** [comment]
sorry, it isn't quite finished

File: lib/forcegraph.js [hunk 2]
```diff
@@ -682,6 +697,8 @@ define(["d3"], function (d3) {
       })
 
       nodes = intNodes.filter(function (d) { return d.o.node })
+      uplinkNodes = intNodes.filter(function (d) { return d.o.node && d.o.node.flags.uplink })
```
Comment (lost-tty) [inline lib/forcegraph.js]: This can be reduced to `nodes.filter(function (d) { return d.o.node.flags.uplink })`

**Comment by lost-tty** [comment]
Could you take a look at the Travis CI output? It looks like there's something wrong.

**Comment by ghost** [comment]
Yes, it's broken now. I'll notify you as soon as it is working again and you can take a look. 

How do we call nodes which aren't uplinks?

What a pity you aren't at 32C3..

**Comment by lost-tty** [comment]
Well, we could still call them nodes. Actually, we could just draw uplink nodes again ontop of the nodes (i.e. draw them twice, but without the shadow). This could make the code a lot simpler and allow other indications for uplinks later. Or we could draw something below them so the hole in the "donut" won't be black but some color.

**Comment by ghost** [comment]
I'm working on it. 

"Der Techniker ist informiert"

**Comment by ghost** [comment]
should be fine now
the blue nodes are a bit dominant now as the 'hotspots' are blue too
also is the shadow rendered twice now?

**Comment by lost-tty** [comment]
I just took a deeper look at the way you're drawing nodes. This is completely broken. `node` and `nctx` is supposed to be a static image of a normal node that's blitted wherever a node should appear. That means, you shouldn't continue drawing ontop of `nctx` after around line 346. Also, you shouldn't reset the scale after that, either. I'm now at #32c3.

**Comment by ghost** [comment]
These changes should eliminate nearly all the things. 
Why should nctx not be changed? It will be rerendered for the next frame anyways?!

**Comment by ghost** [comment]
unfortunately we can't move the *Nodes.filter(visibleNodes) to the bottom, because visibleNodes(d) uses variables that can change every frame. So the filtering has to be in the redraw method.

**Comment by ghost** [comment]
What's up with this now? If the uplink flag doesn't exist the nodes just won't be colored. So I think we can merge this before ffnord/ffmap-backend gets the uplink flag feature.

**Comment by lost-tty** [comment]
Could you rename `nonUplinkNodes` back to `nodes`? I'd also like to see a patch for ffmap-backend first so we can both test and use this feature in the next meshviewer release.

**Comment by ghost** [comment]
1) done.
2) I won't do anything there.

**Comment by ghost** [comment]
build fixed

**Comment by lost-tty** [comment]
Manually merged (and tweaked a little).

**Pull request closed without merging**